Add productType property as an ansible fact

srjennings opened the pull request:
Suggest to add productType property from win32_operatingsystem CIM instance to differentiate between versions and add new fact.

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Sometimes it is necessary to differentiate between Servers, Workstations and Domain Controllers. Adding the ProductType property from the win32_operatingsystem class can help provide ansible the ability to use this property when running certain playbooks or commands and scope accordingly.

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
win_setup

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

Source: https://msdn.microsoft.com/en-us/library/aa394239(v=vs.85).aspx
Value Legend:

* '1' indicates Work Station
*  '2' indicates Domain Controller
*  '3' indicates Server

Commit:
Add productType property as an ansible fact

Suggest to add productType property from win32_operatingsystem CIM instance to differentiate between versions and add new fact.

nitzmahone reviewed (changes requested):
This should be self-describing... eg, do the mapping to a human-readable string [`workstation`,`server`,`domain_controller`,`unknown`]. That last value is just in case we get a null or misunderstood value back (it's WMI, you can't always assume you'll get a usable value)...

Commit:
update code to display human-readable string

Commit added in response to nizmahone's suggestion to add humane-readable strings to the display. uses switch statement to evaluate and update $ansible_facts hash table with proper information

deyjcode commented:
Commit added in response to @nitzmahone  suggestion to add humane-readable strings to the display. Uses switch statement to evaluate and update $ansible_facts hash table with proper information

nitzmahone commented on lib/ansible/modules/windows/setup.ps1:
```diff
@@ -199,6 +199,23 @@ if($gather_subset.Contains('distribution')) {
         ansible_os_family = "Windows"
         ansible_os_name = ($win32_os.Name.Split('|')[0]).Trim()
     }
+    switch ($win32_os.ProductType.ToString()) {
```
This will bomb setup if ProductType returns $null (which I've seen before, believe it or not). The string conversion also adds unnecessary complexity (unless there's a reason for it I'm missing?). How about something like:
```
$pt = switch($win32_os.ProductType) {
    1 { "workstation" }
    2 { "domain_controller" }
    3 { "server" }
    default { "unknown" }
}

$ansible_facts += @{ ansible_os_producttype = $pt }
```

That will properly handle the `$null` case as well as pretty much anything else out of the ordinary, short of an eval exception or the property being missing entirely (neither of which we attempt to handle anywhere else right now).

srjennings replied:
I like that. I was looking at null value but wasn't certain how null was handled in the switch statement.

Commit:
remove string conversion, modify switch default

Adjusts line 202 to remove .String() conversion from uInt32.
Adjusts switch test values to integers
Uses switch default clause to handle null values

deyjcode commented:
modified switch statement

jborean93 commented on lib/ansible/modules/windows/setup.ps1:
```diff
@@ -199,6 +199,20 @@ if($gather_subset.Contains('distribution')) {
         ansible_os_family = "Windows"
         ansible_os_name = ($win32_os.Name.Split('|')[0]).Trim()
     }
+    switch ($win32_os.ProductType) {
```
3 things
* The indenting and formatting are a bit off, you can easily simplify this with the below
* The fact should really be called `ansible_os_product_type` not `ansible_os_producttype`
* Move this before adding the keys above so that operation is only done once and all the `distribution` keys are added together. This makes it easier to see the grouping and maintainability in the future

```
$product_type = switch($win32_os.ProductType) {
    1 { "workstation" }
    2 { "domain_controller" }
    3 { "server" }
    default { "unknown" }
}
$ansible_facts += @{
    ansible_distribution = $win32_os.Caption
    ...  # the other keys here
    ansible_os_product_type = $product_type
}
```

Commit:
formatting

fixed formatting issues, moved product_type into variable.
adjust ansible_fact hash table key to be consistent with the rest of the code.

deyjcode commented:
Changed made. Thanks. Apologies for any frustration here. First time contributing. 

Learned something new with this process. Going to go back into my own code and re-evaluate some decisions I've made to make my own scripts easier to read.

jborean93 committed:
Moved product type within distribution

jborean93 commented:
No worries @deyjcode appreciate you working on this. Just committed a minor change and will merge after tests have run
